### PR TITLE
Add join codes

### PR DESCRIPTION
**Pull request by floogulinc**

### CODE REVIEW

**Comment by helloworld12321** [comment]
Sorry about the force-push. I made a few commits under the wrong git account—that's my fault.

**Comment by floogulinc** [comment]
We should have an error snackbar for when generating a joincode fails (after the retries).

**Comment by helloworld12321** [comment]
> We should have an error snackbar for when generating a joincode fails (after the retries).

Oh, totally, I'll go do that—thank you!

**Comment by helloworld12321** [comment]
```ts
    }, err => {
      // blah blah snackbar
      this.joinCodeButtonDisabled$.next(false);
    });
```

Yeah, that's probably not what we want in production.

**Comment by floogulinc** [comment]
> ```ts
>     }, err => {
>       // blah blah snackbar
>       this.joinCodeButtonDisabled$.next(false);
>     });
> ```
> 
> Yeah, that's probably not what we want in production.

Haha I totally forgot we left that. oops!

**Comment by helloworld12321** [comment]
Cool, should be all done now!

File: src/app/join-code.ts [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+export function joinCodesAreEqual(
+  first: JoinCodeWithId | null,
+  second: JoinCodeWithId | null,
+): boolean {
+  if (first === null && second === null) {
+    return true;
+  } else if (first === null && second !== null) {
+    return false;
+  } else if (first !== null && second === null) {
+    return false;
```
Comment (floogulinc) [inline src/app/join-code.ts:27]: Could this be
```suggestion
  if (first === null || second === null) {
    return first === second;
```

File: src/app/pages/join-session/join-session.component.html [hunk 2]
```diff
@@ -17,7 +20,9 @@
           <input matInput inputmode="numeric" pattern="[0-9]*" class="nest" formControlName="nestControl" required>
           <mat-hint>The number at the top of your nest marker</mat-hint>
           <mat-error *ngIf="sessionFormGroup.controls['nestControl'].invalid">
-            Invalid nest ID: please enter the ID number on your nest marker
+            Invalid nest ID: your input isn't a number in the range
+            {{MIN_NEST_MARKER}}&ndash;{{MAX_NEST_MARKER}}. Please enter the ID
+            number on your nest marker.
```
Comment (floogulinc) [inline src/app/pages/join-session/join-session.component.html:25]: How about
```suggestion
            Invalid nest ID: join codes are numbers between 
            {{MIN_NEST_MARKER}} and {{MAX_NEST_MARKER}}. Please enter the ID
            number on your nest marker.
```

File: src/app/pages/session-lobby/session-lobby.component.ts [hunk 3]
```diff
@@ -51,4 +55,22 @@ export class SessionLobbyComponent implements OnInit {
...
   }
+
+  createJoinCode() {
+    this.joinCodeButtonDisabled$.next(true);
+    this.teacherSessionService.createJoinCode().subscribe(() => {
+      this.joinCodeButtonDisabled$.next(false);
+    }, err => {
+      this.joinCodeButtonDisabled$.next(false);
+      this.matSnackbar.open(
+        `Error: couldn't create a join code. Please try again later.\n(${err})`,
```
Comment (floogulinc) [inline src/app/pages/session-lobby/session-lobby.component.ts:66]: Do we want to include the raw error here?

File: cypress/integration/join-page.spec.ts [hunk 4]
```diff
@@ -130,20 +147,43 @@ describe('The join page', () => {
...
 
       context('Entering bad values into the form', () => {
-        context('Entering a bad session id', () => {
+        describe('Using the malformed join code', () => {
+          for (const badJcId of [
+            'likeZoinksScoob',
+            '1',
+            '1234567',
+            // Eastern Arabic numerals, from Unicode.
+            '٢٢٠٩١٠',
```
Comment (floogulinc) [inline cypress/integration/join-page.spec.ts:156]: A few more fun tests cases:
```suggestion
            'likeZoinksScoob',
            '1',
            '-123456',
            '-12345',
            '1234567',
            '0123456',
            // Eastern Arabic numerals, from Unicode.
            '٢٢٠٩١٠',
            // Chinese/Japanese numerals
            '一二三四五六',
            // Fullwidth characters
            '１２３４５６'
```

File: src/app/pages/join-session/join-session.component.html (hunk 2 above)
Comment (helloworld12321) [reply src/app/pages/join-session/join-session.component.html:25]: Sure—thanks!

File: src/app/join-code.ts (hunk 1 above)
Comment (helloworld12321) [reply src/app/join-code.ts:27]: Ah, clever! Looks good!

File: cypress/integration/join-page.spec.ts (hunk 4 above)
Comment (helloworld12321) [reply cypress/integration/join-page.spec.ts:156]: Nice, thanks :)

File: src/app/pages/session-lobby/session-lobby.component.ts (hunk 3 above)
Comment (helloworld12321) [reply src/app/pages/session-lobby/session-lobby.component.ts:66]: We could, but I'd lean towards not troubling our users with that sort of technical stuff ¯\\\_(ツ)\_/¯
Could go either way though!

**Comment by kklamberty** [comment]
To test this, should I be able to use the Vercel deployment or must I run it locally?

File: src/app/pages/session-lobby/session-lobby.component.ts (hunk 3 above)
Comment (floogulinc) [reply src/app/pages/session-lobby/session-lobby.component.ts:66]: Right, it currently *is* printing the error though.

Comment (helloworld12321) [reply src/app/pages/session-lobby/session-lobby.component.ts]: Oh, sorry :rofl: my bad there :)

**Review by kklamberty — APPROVED** [review]
I noticed that I could run into a time when the code should be valid but it isn't valid. I think this is only in cases where it's really soon after a code was deleted and a new code was generated. Perhaps my 'player/participant' browser needed to be refreshed? Other than that, this all looked good. I think it is not likely to come up in practice.

File: cypress/integration/join-page.spec.ts [hunk 5]
```diff
@@ -146,23 +146,66 @@ describe('The join page', () => {
...
+          mockJoinCodeId.replace(/\B(?=([0-9]{3})+(?![0-9]))/g, ' '),
+          // NO-BREAK SPACE
+          mockJoinCodeId.replace(/\B(?=([0-9]{3})+(?![0-9]))/g, '\u00A0'),
+          // CHARACTER TABULATION
+          mockJoinCodeId.replace(/\B(?=([0-9]{3})+(?![0-9]))/g, '\u0009'),
+          // ZERO WIDTH NO-BREAK SPACE
+          mockJoinCodeId.replace(/\B(?=([0-9]{3})+(?![0-9]))/g, '\uFEFF'),
+          mockJoinCodeId.replace(/\B(?=([0-9])+(?![0-9]))/g, ' '),
+          `   ${mockJoinCodeId.replace(/\B(?=([0-9]{2})+(?![0-9]))/g, '   ')}   `,
+        ]) {
```
Comment (helloworld12321) [inline cypress/integration/join-page.spec.ts:160]: Those are quite the regular expressions! :joy: I appreciate the fanciness, but do you think we could do something like …? (It's more boring, to be sure, but this way I don't have to remember what a negative lookbehind does :stuck_out_tongue_winking_eye:)

```suggestion
        for (const goodJcId of [
          mockJoinCodeId.match(/[0-9]{3}/g).join(' '),
          // NO-BREAK SPACE
          mockJoinCodeId.match(/[0-9]{3}/g).join('\u00A0'),
          // CHARACTER TABULATION
          mockJoinCodeId.match(/[0-9]{3}/g).join('\u0009'),
          // ZERO WIDTH NO-BREAK SPACE
          mockJoinCodeId.match(/[0-9]{3}/g).join('\uFEFF'),
          mockJoinCodeId.split('').join(' '),
          `   ${mockJoinCodeId.match(/[0-9]{2}/g).join('   ')}   `,
        ]) {
```

**Review by helloworld12321 — APPROVED** [review]

**Comment by helloworld12321** [comment]
@kklamberty Oh, funny—are you saying that a student couldn't join the session using the new code?

**Review by UtkarshKr007 — APPROVED** [review]
Couldn't get any undesired results past the join code with the new validation!! Perfect.
The codes did expire for me after an hour, so that's good. Great work on optimazations and refactoring the firestore tests.

**Review by kklamberty — APPROVED** [review]
I used my phone and computer as participants and hosted on my computer. I tried using join codes that had been removed (failed as it should), different codes for each participant (removing code in between), and using a join code for the same session more than one time after closing the participant view and using the back button. It seems to work on all counts, and I think the interface looks great. It allows you to enter your name when you join again, so even without renaming working you can rename yourself if you do it wrong the first time. Nice work!

**Pull request merged**